pkg/quarterdeck/db/models/users.go

```diff
@@ -819,6 +812,83 @@ func (u *User) SwitchOrganization(ctx context.Context, orgID any) (err error) {
 	return tx.Commit()
 }
 
+const (
+	deleteUserOrgSQL           = "DELETE FROM organization_users WHERE user_id=:userID AND organization_id=:orgID"
+	deleteUserInviteByEmailSQL = "DELETE FROM user_invitations WHERE email=:email AND organization_id=:orgID"
+)
+
+// RemoveOrganization removes the user from the specified organization. If this results
+// in the user having no organizations then the user is deleted. This also deletes all
+// invitations that have been sent to the user for the organization and all api keys
+// created by the user.
+func (u *User) RemoveOrganization(ctx context.Context, orgID any) (err error) {
+	var userOrg ulid.ULID
+	if userOrg, err = ulids.Parse(orgID); err != nil {
+		return err
+	}
+
+	if ulids.IsZero(userOrg) {
+		return ErrMissingOrgID
+	}
+
+	// Delete the rest of the organization user resources in a transaction
+	var tx *sql.Tx
+	if tx, err = db.BeginTx(ctx, nil); err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	// Revoke all the keys the user created for this organization
+	if err = u.revokeKeys(tx, userOrg); err != nil {
+		return err
+	}
+
+	// Delete the organization user mapping
+	if _, err = tx.Exec(deleteUserOrgSQL, sql.Named("userID", u.ID), sql.Named("orgID", userOrg)); err != nil {
+		return err
+	}
+
+	// Delete all invitations for the user in the organization
+	if _, err = tx.Exec(deleteUserInviteByEmailSQL, sql.Named("email", u.Email), sql.Named("orgID", userOrg)); err != nil {
+		return err
+	}
+
+	// If the user doesn't have any organizations then delete the user
+	if _, err = u.defaultOrganization(tx); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			if err = u.delete(tx); err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	return tx.Commit()
+}
+
+const (
+	insertRevokedUserKeysSQL = "INSERT INTO revoked_api_keys SELECT k.id, k.key_id, k.name, k.organization_id, k.project_id, k.created_by, k.source, k.user_agent, k.last_used, p.perms, k.created, k.modified FROM api_keys k JOIN (SELECT id, api_key_id, json_group_array(name) AS perms FROM api_key_permissions akp JOIN permissions p ON p.id=akp.permission_id GROUP BY api_key_id) p ON p.api_key_id=k.id WHERE k.created_by=:userID AND k.organization_id=:orgID"
```

Comment:
Very nicely done; I didn't know about json_group_array -- I'll have to look into this one. 

I did some brief research into sqlite3 UDFs -- it may be possible but if this is well tested I'm happy to move forward with it for now. Maybe we could break the constant up a bit like this (for maintainability):

```suggestion
	insertRevokedUserKeysSQL = `INSERT INTO revoked_api_keys 
		SELECT k.id, k.key_id, k.name, k.organization_id, k.project_id, k.created_by, k.source, k.user_agent, k.last_used, p.perms, k.created, k.modified
		FROM api_keys k 
		JOIN (
			SELECT id, api_key_id, json_group_array(name) AS perms 
				FROM api_key_permissions akp 
				JOIN permissions p ON p.id=akp.permission_id 
			GROUP BY api_key_id
		) p ON p.api_key_id=k.id 
		WHERE k.created_by=:userID AND k.organization_id=:orgID`
```

The `` allows you to create multiline string constants.